### DISCUSSION TITLE
Number of cores used in simulation

### DISCUSSION THREAD

**Discussion by MathXCruz** [microcombustion/ctwrap#89] [Q&A] [answered]
Hello,

Is there a way I can specify the number of cores I want to use in a simulation? Looking at handler.py I see the standard is number_of_cores // 2, but my freeflame.py simulations only use 1 core in my 4 cores CPU.

**Comment by ischoegl** [comment, marked as answer]
Hi, could you post what command you used? (Especially what options you specified)

**Reply by MathXCruz** [reply]
The command I used was

`ctwrap run freeflame freeflame.yaml --parallel --strategy matrix`

**Reply by ischoegl** [reply]
Could you add the -v option and report what you see (there is also -vv if I recall)

**Reply by MathXCruz** [reply]
This is what I got from running with -vv. I interrupted the simulation to save some time. It seems it's working with multiple parallel processes and I could have misinterpreted computer resource monitoring information.

```
################################################################################
Running simulations: module `ctwrap.modules.freeflame`
Conditions specified in `freeflame.yaml` 
################################################################################
Simulations for entries ['upstream.T', 'upstream.P', 'model.transport']
 * Starting parallel batch simulation using 4 cores
   - starting Process-1
   - starting Process-2
 * running `case_00` (Process-1)
   - rescheduling case_01 (Process-2)
 * running `case_02` (Process-2)
   - starting Process-3
   - starting Process-4
   - rescheduling case_03 (Process-3)
   - terminating Process-4
 * running `case_04` (Process-3)
    flamespeed = 0.134950 m/s (mix)
   - rescheduling case_05 (Process-2)
 * running `case_06` (Process-2)
    flamespeed = 0.100344 m/s (mix)
   - rescheduling case_07 (Process-3)
 * running `case_08` (Process-3)
    flamespeed = 0.463403 m/s (mix)
   - rescheduling case_09 (Process-1)
 * running `case_10` (Process-1)
    flamespeed = 1.227211 m/s (mix)
   - rescheduling case_11 (Process-3)
 * running `case_12` (Process-3)
    flamespeed = 0.416065 m/s (mix)
   - rescheduling case_13 (Process-1)
 * running `case_14` (Process-1)
    flamespeed = 0.085060 m/s (mix)
   - rescheduling case_15 (Process-2)
 * running `case_16` (Process-2)
    flamespeed = 2.523672 m/s (mix)
 * running `case_17` (Process-2)
    flamespeed = 0.311124 m/s (mix)
 * running `case_18` (Process-3)
    flamespeed = 0.265700 m/s (mix)
   - rescheduling case_19 (Process-1)
 * running `case_20` (Process-1)
    flamespeed = 0.987952 m/s (mix)
   - rescheduling case_21 (Process-3)
 * running `case_22` (Process-3)
    flamespeed = 0.756968 m/s (mix)
   - rescheduling case_23 (Process-1)
 * running `case_01` (Process-1)
    flamespeed = 0.646113 m/s (mix)
 * running `case_03` (Process-3)
    flamespeed = 2.491844 m/s (soret)
 * running `case_05` (Process-2)
    flamespeed = 0.459130 m/s (soret)
 * running `case_07` (Process-1)

**Reply by ischoegl** [reply]
Ok. That’s interesting - from my perspective, I am relying on Python’s libraries for multiprocessing, but cannot rule out that there’s something awry. I recall this working on my Linux machine. What OS are you running?

**Reply by MathXCruz** [reply]
I'm using Ubuntu 20.04.

**Reply by ischoegl** [reply]
I just verified on my Fedora 32 installation that multiple cores become active - I believe you should have `System Monitor` also (which is what I used). Would post a screenshot if my Gnome PrtScr weren't crashing ... If you only see a single process, please let me know.

**Reply by MathXCruz** [reply]
I can see it working as expected now. The problem was probably caused by a simulation with too few cases to use all the cores at the same time. Thank you for your help!

**Comment by ischoegl** [comment]
On a separate note, the command line interface currently doesn't support the specification of a number of cores. However, you can use the scripted API ... see [minimal example](https://microcombustion.github.io/ctwrap/examples/minimal_example.html#id2), where you can specify the option `number_of_processes` in [`run_parallel`](https://microcombustion.github.io/ctwrap/pages/handler.html#ctwrap.handler.SimulationHandler.run_parallel).
